pkg/processor/metricfrequencyprocessor/strategy.go

```diff
@@ -0,0 +1,182 @@
+	"math"
+	"sort"
+	"time"
+)
+
+const (
+	float64EqualityThreshold = 1e-9
+	minPointAccumulationTime = time.Minute * 15
+	category1ReportFrequency = time.Second * 30
+	category2ReportFrequency = time.Minute * 2
+	category3ReportFrequency = time.Minute * 5
+	safetyInterval           = time.Second * 1
+
+	iqrAnomalyCoef            = 1.5
+	variationIqrThresholdCoef = 4
+)
+
+type MetricSieve struct {
+	metricCache  MetricCache
+	lastReported map[string]pdata.Timestamp
+}
+
+func (fs *MetricSieve) Sift(metric pdata.Metric) bool {
+	if metric.DataType() == pdata.MetricDataTypeDoubleGauge {
+		return fs.siftDropGauge(metric)
+	} else {
+		return false
+	}
+}
+
+func (fs *MetricSieve) siftDropGauge(metric pdata.Metric) bool {
+	metric.DoubleGauge().DataPoints().RemoveIf(fs.siftDataPoint(metric.Name()))
+
+	return metric.DoubleGauge().DataPoints().Len() == 0
+}
+
+func (fs *MetricSieve) siftDataPoint(name string) func(pdata.DoubleDataPoint) bool {
+	return func(dataPoint pdata.DoubleDataPoint) bool {
+		cachedPoints := fs.metricCache.List(name)
+		fs.metricCache.Register(name, dataPoint)
+		lastReported, _ := fs.lastReported[name]
+		earliest := earliestTimestamp(cachedPoints)
+		cachedPoints[dataPoint.Timestamp()] = dataPoint.Value()
+
+		if metricRequiresSamples(dataPoint, earliest) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category3ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if isConstant(dataPoint, cachedPoints) {
+			return true
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category2ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if isLowInformation(cachedPoints) {
+			return true
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category1ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		return true
+	}
+}
+
+func metricRequiresSamples(point pdata.DoubleDataPoint, earliest pdata.Timestamp) bool {
+	return point.Timestamp().AsTime().Before(earliest.AsTime().Add(minPointAccumulationTime))
+}
+
+func pastCategoryFrequency(point pdata.DoubleDataPoint, lastReport pdata.Timestamp, categoryFrequency time.Duration) bool {
+	return point.Timestamp().AsTime().After(lastReport.AsTime().Add(categoryFrequency).Add(safetyInterval))
+}
+
+func isConstant(point pdata.DoubleDataPoint, points map[pdata.Timestamp]float64) bool {
+	for _, value := range points {
+		if !almostEqual(point.Value(), value) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isLowInformation(points map[pdata.Timestamp]float64) bool {
+	q1, q3 := calculateQ1Q3(points)
+	iqr := q3 - q1
+	variation := calculateVariation(points)
+
+	return withingBounds(points, q1-iqrAnomalyCoef*iqr, q3+iqrAnomalyCoef*iqr) && lowVariation(variation, iqr)
+}
+
+func calculateQ1Q3(points map[pdata.Timestamp]float64) (float64, float64) {
```

Comment:
Can you sprinkle some more comment around funcs like this?

Reply:
Added few comments.